### DISCUSSION TITLE
How does pydantic validates that dict has a key ?

### DISCUSSION THREAD

**Discussion by vitalik** [pydantic/pydantic#7584] [Pydantic V2]
I'm trying to get some extra behaviour out of dict validation:

```Python
from pydantic import BaseModel

class MyDict(dict):
    def __getitem__(self, key: str):
        print(" * calling __getitem__", key)
        return super().__getitem__(key)

    def __contains__(self, key):
        print(f" * calling __contains__ {key}")
        return super().__contains__(key)

    def keys(self):
        print(f" * calling keys")
        return super().keys()

    def get(self, key):
        print(f" * calling get({key})")
        return super().get(key)

d = MyDict({"x": 1})

class Data(BaseModel):
    x: int

validated = Data.model_validate(d).model_dump()
```

but I do not see any of the calls in stdout... how exactly pydantic accessing dict values or check if they exist in `extra=forbid` mode ?

**Comment by samuelcolvin** [comment]
if `extra=forbid` or `extra=allow` we iterate over the dict after looking up fields, see [here](https://github.com/pydantic/pydantic-core/blob/91746c9565c43739c7a8f392209bb6d1aeb06d8a/src/validators/model_fields.rs#L233), and [here](https://github.com/pydantic/pydantic-core/blob/91746c9565c43739c7a8f392209bb6d1aeb06d8a/src/input/return_enums.rs#L443-L460) for the implementation of `DictGenericIterator`.

It's most likely `items()` that you need to instrument, but I've no idea if PyO3's `PyDictIterator` will actually call your subclass implementation of `items()`.

We need the value as well as the key to populate the error as well as for the case of `extra=allow`.

**Reply by samuelcolvin** [reply]
This code works:

```py
from collections import UserDict

from pydantic import BaseModel

class MyDict(UserDict):
    def __getitem__(self, key: str):
        print(" * calling __getitem__", key)
        return super().__getitem__(key)

    def items(self):
        print(f" * calling items")
        return super().items()

d = MyDict({"x": 1})

class Data(BaseModel, extra='forbid'):
    x: int

validated = Data.model_validate(d).model_dump()
```
outputs:
```
 * calling __getitem__ x
 * calling items
 * calling __getitem__ x
```

**Reply by vitalik** [reply]
Got it, Thank you
